core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java

```diff
@@ -2407,45 +2410,51 @@ private void loadMetadataInternal(RpcContext rpcContext, LockedInodePath inodePa
         loadDirectoryMetadata(rpcContext, inodePath, context);
 
         if (context.getOptions().getLoadDescendantType() != LoadDescendantPType.NONE) {
-          ListOptions listOptions = ListOptions.defaults();
-          if (context.getOptions().getLoadDescendantType() == LoadDescendantPType.ALL) {
-            listOptions.setRecursive(true);
+          UfsStatus[] children;
+          if (statusCache != null
+              && context.getOptions().getLoadDescendantType() == LoadDescendantPType.ALL) {
+            // Avoid re-caching. Makes the assumption the cache was created on the same ufs URI
```

Comment:
Does this assume the status cache must have been created with ALL (or recursive)? Is that a valid assumption?

Also, what does "avoid re-caching" mean? I don't see where the caching is happening.

Reply:
There are only two callers of `populateStatusCache`, both of which specify `ALL`.

I can add a parameter to the status cache which can identify the load type that the cache was created with. If it's not `ALL` and is not equal to the current LoadDescendantType, then I would probably just forgo using the cache. Most performance gain is only going to come when we have a huge tree after doing a recursive listing (called with `ALL`).

Reply:
"re-caching" is an error. It should probably say "avoid retrieving statuses again"

Reply:
updated